sys/drv_stdvga.c

```diff
@@ -170,9 +171,13 @@ static int stdvga_attach(device_t *dev) {
   pci_write_config(dev, PCIR_COMMAND, 2, command);
 
   stdvga_state_t *stdvga = dev->state;
-  /* TODO: This will get replaced by bus_alloc_resource* function */
-  stdvga->mem = &pcid->bar[0];
-  stdvga->io = &pcid->bar[1];
+
+  stdvga->mem = bus_resource_alloc_anywhere(
+    dev, pcid->bar_info[0].type, PCIR_BAR(pcid->bar_info[0].rid),
```

Comment:
Using `pcid->bar_info[0].type` here is cheating. In this case the driver should explicitely request BAR[0] as memory resource, and BAR[2] (sic!) as I/O space.

Reply:
Why is it cheating? `bar_info` belongs to this device. Are you suggesting that device should pass only bar number to `bus_resource_alloc` and bus driver should get resource info from `bar_info` ?

Reply:
I agree that it is ugly. This driver should rather search `bar_info` array for non-empty bars.

Reply:
BAR numbers used by a driver should not depend on particular method of BAR discovery. If a device facilitate `BAR[0]` and `BAR[2]` one should be able to just pass `rid=0` and `rid=2` (respectively) to `bus_resource_alloc*`.

I'm expecting following two lines to look as follows:
```c
stdvga->mem = bus_resource_alloc_anywhere(dev, RT_MEMORY, 0, ~0, RT_PREFETCHABLE);
stdvga->io = bus_resource_alloc_anywhere(dev, RT_IOPORTS, 2, ~0, 0);
```